File: services/server/src/safe_socket/safe_socket.py
import socket

# TODO: Complete with a short-read/short-write tolerant implementation


def recv_all(socket: socket.socket, size):
    buffer = bytearray()

    while len(buffer) < size:
        chunck = socket.recv(size - len(buffer))
        if not chunck:
            raise ConnectionError (
                "Conection closes by peer before all data was received"
            )
        buffer.extend(chunck)

    return bytes(buffer)



def send_all(socket: socket.socket, bytes):
    total_sent = 0
    while total_sent < len(bytes):
        sent = socket.send(bytes[total_sent:])
        total_sent += sent

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"

//TODO: Complete with a short-read/short-write tolerant implementation

func SendAll(writer io.Writer, data []byte) error {
	totalWritten := 0
	for totalWritten < len(data) {
		n, err := writer.Write(data[totalWritten:])
		totalWritten += n
		if err != nil {
			return err
		}
	}
	return nil
}

func RecvAll(reader io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	totalRead := 0
	for totalRead < size {
		n, err := reader.Read(buff[totalRead:])
		totalRead += n
		if err != nil {
			if err == io.EOF && totalRead == size {
				break
			}
			return nil, err
		}
	}
	return buff, nil
}
